xdem/coreg.py

```diff
@@ -1063,3 +1081,620 @@ def coregister(reference_raster: Union[str, gu.georaster.Raster], to_be_aligned_
+            args=(x_coords, y_coords, ddem),
+            disp=False
+        )
+
+        self._meta["coefficients"] = coefs
+        self._meta["func"] = lambda x, y: poly2d(x, y, coefs)
+
+    def _apply_func(self, dem: np.ndarray, transform: rio.transform.Affine) -> np.ndarray:
+        """Apply the deramp function to a DEM."""
+        bounds, resolution = _transform_to_bounds_and_res(dem, transform)
+        x_coords, y_coords = np.meshgrid(
+            np.linspace(bounds.left + resolution / 2, bounds.right - resolution / 2, num=dem.shape[1]),
+            np.linspace(bounds.bottom + resolution / 2, bounds.top - resolution / 2, num=dem.shape[0])[::-1]
+        )
+
+        ramp = self._meta["func"](x_coords, y_coords)
+
+        return dem + ramp
+
+    def _apply_pts_func(self, coords: np.ndarray) -> np.ndarray:
+        """Apply the deramp function to a set of points."""
+        new_coords = coords.copy()
+
+        new_coords[:, 2] += self._meta["func"](new_coords[:, 0], new_coords[:, 1])
+
+        return new_coords
+
+    def _to_matrix_func(self) -> np.ndarray:
+        """Return a transform matrix if possible."""
+        if self.degree > 1:
+            raise ValueError(
+                "Nonlinear deramping degrees cannot be represented as transformation matrices."
+                f" (max 1, given: {self.degree})")
+        if self.degree == 1:
+            raise NotImplementedError("Vertical shift, rotation and horizontal scaling has to be implemented.")
+
+        # If degree==0, it's just a bias correction
+        empty_matrix = np.array([[1, 0, 0, 0], [0, 1, 0, 0], [0, 0, 1, 0], [0, 0, 0, 1]], dtype=float)
+
+        empty_matrix[2, 3] += self._meta["coefficients"][0]
+
+        return empty_matrix
+
+
+class CoregPipeline(Coreg):
+    """
+    A sequential set of coregistration steps.
+    """
+
+    def __init__(self, pipeline: list[Coreg]):  # pylint: disable=super-init-not-called
+        """
+        Instantiate a new coregistration pipeline.
+
+        :param: Coregistration steps to run in the sequence they are given.
+        """
+        self.pipeline = pipeline
+        self._meta = {}
+
+    def _fit_func(self, ref_dem: np.ndarray, tba_dem: np.ndarray, transform: Optional[rio.transform.Affine],
+                  weights: Optional[np.ndarray]):
+        """Fit each coregistration step with the previously transformed DEM."""
+        tba_dem_mod = tba_dem.copy()
+
+        for coreg in self.pipeline:
+            coreg._fit_func(ref_dem, tba_dem_mod, transform=transform, weights=weights)
+            coreg._fit_called = True
+
+            tba_dem_mod = coreg._apply_func(tba_dem_mod, transform)
+
+    def _apply_func(self, dem: np.ndarray, transform: rio.transform.Affine) -> np.ndarray:
+        """Apply the coregistration steps sequentially to a DEM."""
+        dem_mod = dem.copy()
+
+        for coreg in self.pipeline:
+            dem_mod = coreg.apply(dem_mod, transform)
```

Comment:
That's really fine for now, but in the future, I think this could be done in one step using some matrix magic! ;-)

Reply:
Indeed, I can add an issue for it.

Problems arise with intermediate non-matrixable steps such as Deramp. It could be implemented in a way where all "matrixable" transforms that occur in a row are merged:

pipeline = [A, B, C (matrix not supported), D, E, F]

apply:
A-B, C, D-F

Reply:
Yes, but as discussed somewhere else with @rhugonnet, I think all transformations that cannot be converted into a matrix format should go into biascorr.py (this includes deramp) and should be handled a bit differently. So I wouldn't worry to much about deramp here.